docs_src/design/ucr/Global Common Configuration.md

```diff
@@ -0,0 +1,169 @@
+
+There are other similar common sections not shown shown above. As can be seen from the two examples above there is much duplication of configuration settings across all the EdgeX services. This gives rise to the need to have all these common duplicate configuration settings in a single global source. 
+
+### Existing solutions
+There are no existing solutions for global configuration for EdgeX since the current configuration implementation is specific to EdgeX.
+
+### Requirements
+
+#### General
+
+- Services shall be able reference a global common configuration in a manner that is flexible for use with and without the Service Provider
+- Common environment overrides shall be applied to global common configuration settings only once before the services receive the setting values
+
+    - This avoids the need for duplicate environment overrides on each service in the docker compose files
+
+- Services must be able to override any of the global common configuration settings with local service specific TOML values
+
+    !!! example - "Example Core Data specific **Writable.Telemetry** and **Service** configuration settings in local TOML file"
+        ```toml
+          [Writable.Telemetry]
+            [Writable.Telemetry.Metrics] # All service's metric names must be present in this list.
+            EventsPersisted = false
+            ReadingsPersisted = false
+
+        ...
+        
+        [Service]
+        Port = 59880
+        StartupMsg = "This is the Core Data Microservice"
+        ```
+
+- Environment variable overrides shall be applied in the same manner as they are today. 
+    - After loaded from file (prior to pushing into the the Configuration Provider) 
+    - After loaded from the Configuration Provider
+
+#### With Configuration Provider
+
+- Global common configuration shall be pre-loaded into the Configuration Provider where it is pulled from by each service.
+    - How and where the configuration is loaded into the Configuration Provider is left to design
+- Pre-loading of global common configuration must have a mechanism to indicate that the configuration pre-load is complete 
+- Each service must wait until the global common configuration pre-load is complete.
+- Each service's full configuration, global common and local TOML overrides, shall be pushed into the Configuration Provider
+    - Once initially loaded the service's full configuration is solely referenced from the Configuration Provider as it is today
+- The services shall be notified when a setting in the **Writeable** section of global common configuration has changed. Each service shall consume the updated setting into its on `Writable` section and push the update back into the Configuration Provider under the service's own full configuration

```

Comment:
What if the service's own entry for such a setting has been customized, and subsequently the global configuration is changed?
I'd simplify the requirement so that the only settings held globally are those which cannot reasonably be different between services, eg the location of core-metadata, the internal messagebus settings. These settings would not then be pushed into the individual service's configuration, but be held only in the global config.

Reply:
So what I hear from this is to rework the service's configuration so that there is totally separate Global Common config and a service specific section. These would be totally separated so that the Global Common is not included in what the service's push into Consul. 

Example for Service:
## Global Common
```toml
[Common.Service]
HealthCheckInterval = '10s'
# MaxRequestSize limit the request body size in byte of put command
MaxRequestSize = 0 # value 0 unlimit the request size.
RequestTimeout = '5s'
  [Service.CORSConfiguration]
  EnableCORS = false
  CORSAllowCredentials = false
  CORSAllowedOrigin = "https://localhost"
  CORSAllowedMethods = "GET, POST, PUT, PATCH, DELETE"
  CORSAllowedHeaders = "Authorization, Accept, Accept-Language, Content-Language, Content-Type, X-Correlation-ID"
  CORSExposeHeaders = "Cache-Control, Content-Language, Content-Length, Content-Type, Expires, Last-Modified, Pragma, X-Correlation-ID"
  CORSMaxAge = 3600

```
## Service Specific
```
[Service]
Host = 'localhost' # must be specific so can be properly overridden in docker compose
Port = 59988
ServerBindAddr = ''  # blank value defaults to Service.Host value  # must be specific so can be properly overridden in docker compose, if needed
StartupMsg = 'some start-up message'
```

This is an interesting idea and warrants more discussion during a meeting.